migration/src/main/scala/akka/persistence/r2dbc/migration/MigrationTool.scala

```diff
@@ -345,4 +369,113 @@ class MigrationTool(system: ActorSystem[_]) {
+          val result = DurableStateResult(acc.persistenceIds + 1, acc.states + states)
+          if (result.persistenceIds % 100 == 0)
+            log.infoN("Migrated [{}] persistenceIds with [{}] durable states.", result.persistenceIds, result.states)
+          result
+        })
+
+    result.transform {
+      case s @ Success(DurableStateResult(persistenceIds, states)) =>
+        log.infoN(
+          "Migration successful. Migrated [{}] persistenceIds with [{}] durable states.",
+          persistenceIds,
+          states)
+        s
+      case f @ Failure(exc) =>
+        log.error("Migration failed.", exc)
+        f
+    }
+  }
+
+  /**
+   * Migrate Durable State for a single persistence id.
+   */
+  def migrateDurableState(persistenceId: String): Future[Int] = {
+    for {
+      _ <- createProgressTable
+      currentProgress <- migrationDao.currentProgress(persistenceId)
+      stateCount <- migrateDurableState(persistenceId, currentProgress)
+    } yield stateCount
+  }
+
+  private def migrateDurableState(persistenceId: String, currentProgress: Option[CurrentProgress]): Future[Int] = {
+    val progressRevision = currentProgress.map(_.durableStateRevision).getOrElse(0L)
+    loadSourceDurableState(persistenceId, progressRevision + 1).flatMap {
+      case None => Future.successful(0)
+      case Some(selectedDurableState) =>
+        for {
+          revision <- {
+            val serializedRow = serializedDurableStateRow(selectedDurableState)
+            targetDurableStateDao
+              .upsertState(serializedRow, selectedDurableState.value, changeEvent = None)
+              .map(_ => selectedDurableState.revision)(ExecutionContexts.parasitic)
+          }
+          _ <- migrationDao.updateDurableStateProgress(persistenceId, revision)
+        } yield 1
+    }
+  }
+
+  private lazy val checkAssertSingleWriter: Unit = {
+    if (targetR2dbcSettings.durableStateAssertSingleWriter) {
+      throw new IllegalArgumentException(
+        "When running the MigrationTool the " +
+        "`akka.persistence.r2dbc.state.assert-single-writer` configuration must be set to off.")
+    }
+  }
+
+  private def serializedDurableStateRow(selectedDurableState: SelectedDurableState): SerializedStateRow = {
+    val stateAnyRef = selectedDurableState.value.asInstanceOf[AnyRef]
+    val serializedState = serialization.serialize(stateAnyRef).get
+    val stateSerializer = serialization.findSerializerFor(stateAnyRef)
+    val stateManifest = Serializers.manifestFor(stateSerializer, stateAnyRef)
+
+    // not possible to preserve timestamp
+    val timestamp = Instant.now()
+
+    val serializedRow = SerializedStateRow(
+      selectedDurableState.persistenceId,
+      selectedDurableState.revision,
+      timestamp,
+      timestamp,
+      Some(serializedState),
+      stateSerializer.identifier,
+      stateManifest,
+      tags = Set.empty // not possible to preserve tags
```

Comment:
Is that a separate table in JDBC or why?

Reply:
we use the the `DurableStateStore` api, so it's not tied to jdbc plugin specifically, and from that we don't get the timestamp

Reply:
I answered above question about timestamp, but same answer for tags